libcalico-go/lib/backend/syncersv1/dedupebuffer/dedupe_buffer.go

```diff
@@ -0,0 +1,283 @@
+			} else {
+				// Update to a key that's already on the queue, swap in the
+				// most recent value.
+				if debug {
+					log.WithField("key", key).Debug("Key updated before being sent.")
+				}
+				usk := element.Value.(updateWithStringKey)
+				usk.update = u
+				element.Value = usk
+			}
+		} else {
+			// No in-flight entry for this key.  Add to queue and record that
+			// it's in flight.
+			if debug {
+				log.WithField("key", key).Debug("No in flight value for key, adding to queue.")
+			}
+			element = d.pendingUpdates.PushBack(updateWithStringKey{
+				key:    key,
+				update: u,
+			})
+			d.keyToPendingUpdate[key] = element
+		}
+	}
+	queueNowEmpty := d.pendingUpdates.Len() == 0
+	if queueWasEmpty && !queueNowEmpty {
+		// Only need to signal when the first item goes on the queue.
+		if debug {
+			log.Debug("Queue transitioned to non-empty; signalling.")
+		}
+		d.cond.Signal()
+	}
+}
+
+func (d *DedupeBuffer) SendToSinkForever(sink api.SyncerCallbacks) {
+	for !d.stopped {
+		d.sendNextBatchToSink(sink)
+	}
+}
+
+func (d *DedupeBuffer) Stop() {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	d.stopped = true
+	d.cond.Signal()
+}
+
+type updateWithStringKey struct {
+	key    string
+	update api.Update
+}
+
+func (d *DedupeBuffer) sendNextBatchToSink(sink api.SyncerCallbacks) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	for d.pendingUpdates.Len() == 0 {
+		if d.stopped {
+			return
+		}
+		d.cond.Wait()
+	}
+	const batchSize = 100
+	buf := make([]any, 0, batchSize)
+	for d.pendingUpdates.Len() > 0 {
+		if d.stopped {
+			return
+		}
+
+		buf = d.pullNextBatch(buf, batchSize)
+		d.dropLockAndSendBatch(sink, buf)
+	}
+}
+
+func (d *DedupeBuffer) pullNextBatch(buf []any, batchSize int) []any {
+	// Grab a batch of updates off the queue.
+	buf = buf[:0]
+	for len(buf) < batchSize && d.pendingUpdates.Len() > 0 {
+		first := d.pendingUpdates.Front()
+		buf = append(buf, first.Value)
+		d.pendingUpdates.Remove(first)
+		if u, ok := first.Value.(updateWithStringKey); ok {
+			key := u.key
+			delete(d.keyToPendingUpdate, key)
+			// Update sentKeys now, before we drop the lock.  Once we drop
+			// the lock we're committed to sending these keys.
+			if u.update.Value == nil {
+				d.sentKeys.Discard(key)
+			} else {
+				d.sentKeys.Add(key)
+			}
+		}
+	}
+	return buf
+}
+
+func (d *DedupeBuffer) dropLockAndSendBatch(sink api.SyncerCallbacks, buf []any) {
+	// RELEASE(!) lock while we send updates downstream.
+	d.lock.Unlock()
+	defer d.lock.Lock()
+
+	debug := log.IsLevelEnabled(log.DebugLevel)
+	updates := make([]api.Update, 0, len(buf))
+	for _, msg := range buf {
+		switch msg := msg.(type) {
+		case updateWithStringKey:
+			updates = append(updates, msg.update)
+		case api.SyncStatus:
+			if len(updates) > 0 {
+				if debug {
+					log.WithField("updates", updates).Debug("Sending updates (pre status update)")
+				}
+				sink.OnUpdates(updates)
+				updates = updates[len(updates):] // Re-slice to end so we don't share storage.
```

Comment:
Does this still preserve capacity?  I guess it probably does, so just checking.

Reply:
Yes, I was pretty sure but I checked the playground: https://go.dev/play/p/7niEE3qk93h